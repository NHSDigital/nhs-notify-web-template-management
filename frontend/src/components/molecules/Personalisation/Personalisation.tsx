import content from '@content/content';
import { ContentRenderer } from '@molecules/ContentRenderer/ContentRenderer';
import { toKebabCase } from '@utils/kebab-case';
import { Details } from 'nhsuk-react-components';

const personalisationContent = content.components.personalisation;

export function Personalisation() {
  return (
    <>
      <h2 className='nhsuk-heading-m' data-testid='personalisation-header'>
        {personalisationContent.header}
      </h2>

      <ContentRenderer content={personalisationContent.leadParagraph}/>

      {personalisationContent.details.map((section) => {
        const id = toKebabCase(section.title);
        return (
        <Details key={id} data-testid={`${id}-details`}>
          <Details.Summary data-testid={`${id}-summary`}>
            {section.title}
          </Details.Summary>
          <Details.Text data-testid={`${id}-text`}>
            <ContentRenderer
              content={section.content}
            />
          </Details.Text>
        </Details>
      )})}
    </>
  );
}
