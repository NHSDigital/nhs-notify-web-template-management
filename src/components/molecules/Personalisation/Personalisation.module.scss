.codeBackground{
  display: inline-block;
  padding: 0.6rem;
  background-color: #d8dde0;
}
