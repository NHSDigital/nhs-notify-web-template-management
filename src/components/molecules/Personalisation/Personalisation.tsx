import content from '@content/content';
import { Details } from 'nhsuk-react-components';

const personalisationContent = content.components.personalisationComponent;

export function Personalisation() {
  return (
    <>
      <h2 className='nhsuk-heading-m' data-testid='personalisation-header'>
        {personalisationContent.header}
      </h2>
      <Details data-testid='personalisation-details'>
        <Details.Summary data-testid='personalisation-summary'>
          {personalisationContent.details.title}
        </Details.Summary>
        <Details.Text data-testid='personalisation-text'>
          <p>{personalisationContent.details.text1}</p>
          <code>{personalisationContent.details.codeBlockText}</code>
          <p className='nhsuk-u-margin-top-4'>
            {personalisationContent.details.text2}
          </p>
          <p>{personalisationContent.details.text3}</p>
          <ul>
            {personalisationContent.details.list.map(({ id, item }) => (
              <li key={id}>{item}</li>
            ))}
          </ul>
        </Details.Text>
      </Details>
    </>
  );
}
