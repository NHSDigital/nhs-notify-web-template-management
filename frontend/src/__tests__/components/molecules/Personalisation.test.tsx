import { render, screen } from '@testing-library/react';
import { Personalisation } from '@molecules/Personalisation/Personalisation';

describe('Personalisation component', () => {
  it('renders component correctly', async () => {
    render(<Personalisation />);

    expect(screen.getByTestId('personalisation-header')).toBeInTheDocument();

    const pdsDetails = screen.getByTestId('pds-personalisation-fields-details');
    expect(pdsDetails).not.toHaveAttribute('open');

    expect(screen.getByTestId('pds-personalisation-fields-summary')).toBeInTheDocument();
    expect(screen.getByTestId('pds-personalisation-fields-text')).toBeInTheDocument();

    const customDetails = screen.getByTestId('custom-personalisation-fields-details');
    expect(customDetails).not.toHaveAttribute('open');

    expect(screen.getByTestId('custom-personalisation-fields-summary')).toBeInTheDocument();
    expect(screen.getByTestId('custom-personalisation-fields-text')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<Personalisation />);
    expect(container).toMatchSnapshot();
  });
});
