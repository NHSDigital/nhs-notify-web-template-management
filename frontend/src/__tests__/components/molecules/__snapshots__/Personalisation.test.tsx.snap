// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Personalisation component matches snapshot 1`] = `
<div>
  <h2
    class="nhsuk-heading-m"
    data-testid="personalisation-header"
  >
    Personalisation
  </h2>
  <p>
    Use double brackets to add a personalisation field to your content.
  </p>
  <p>
    Do not include spaces in your personalisation fields. For example:
  </p>
  <span
    class="nhsuk-u-visually-hidden"
    id="personalisation-markdown-description"
  >
    An example of personalised message content:
  </span>
  <pre
    class="nhsuk-code-block"
  >
    <code
      aria-describedby="personalisation-markdown-description"
    >
      Hello ((firstName)), your NHS number is ((nhsNumber))
    </code>
  </pre>
  <details
    class="nhsuk-details"
    data-testid="pds-personalisation-fields-details"
  >
    <summary
      class="nhsuk-details__summary"
      data-testid="pds-personalisation-fields-summary"
    >
      <span
        class="nhsuk-details__summary-text"
      >
        PDS personalisation fields
      </span>
    </summary>
    <div
      class="nhsuk-details__text"
      data-testid="pds-personalisation-fields-text"
    >
      <p>
        NHS Notify gets data from PDS to populate certain personalisation fields.
      </p>
      <p>
        You can use the following PDS personalisation fields:
      </p>
      <ul>
        <li>
          ((fullName))
        </li>
        <li>
          ((firstName))
        </li>
        <li>
          ((lastName))
        </li>
        <li>
          ((nhsNumber))
        </li>
        <li>
          ((date))
        </li>
      </ul>
      <p>
        Make sure your personalisation fields exactly match the PDS personalisation fields. This includes using the correct order of upper and lower case letters.
      </p>
    </div>
  </details>
  <details
    class="nhsuk-details"
    data-testid="custom-personalisation-fields-details"
  >
    <summary
      class="nhsuk-details__summary"
      data-testid="custom-personalisation-fields-summary"
    >
      <span
        class="nhsuk-details__summary-text"
      >
        Custom personalisation fields
      </span>
    </summary>
    <div
      class="nhsuk-details__text"
      data-testid="custom-personalisation-fields-text"
    >
      <p>
        You can add
        <a
          href="/using-nhs-notify/personalisation#custom-personalisation-fields"
          rel="noopener noreferrer"
          target="_blank"
        >
          custom personalisation fields
        </a>
         that use your own personalisation data.
      </p>
      <p>
        Include custom personalisation fields in your content. Then provide your custom personalisation data using
        <a
          href="/using-nhs-notify/api"
          rel="noopener noreferrer"
          target="_blank"
        >
          NHS Notify API
        </a>
         or
        <a
          href="/using-nhs-notify/mesh"
          rel="noopener noreferrer"
          target="_blank"
        >
          NHS Notify MESH
        </a>
        .
      </p>
      <p>
        For example, if you wanted to include GP surgery as custom personalisation data, your custom personalisation field could be:
      </p>
      <span
        class="nhsuk-u-visually-hidden"
        id="custom-personalisation-markdown-description"
      >
        An example of personalised message content:
      </span>
      <pre
        class="nhsuk-code-block"
      >
        <code
          aria-describedby="custom-personalisation-markdown-description"
        >
          ((GP_surgery))
        </code>
      </pre>
      <p>
        Remember not to include spaces in your personalisation fields.
      </p>
    </div>
  </details>
</div>
`;
